File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/lint/LintProblemReporter.java
package com.github.jmlparser.lint;

import com.github.javaparser.TokenRange;
import com.github.javaparser.ast.nodeTypes.NodeWithTokenRange;

import java.util.function.Consumer;

import static com.github.javaparser.utils.CodeGenerationUtils.f;

/**
 * @author Alexander Weigl
 * @version 1 (13.10.22)
 */
public class LintProblemReporter {
    private final Consumer<LintProblem> problemConsumer;

    public LintProblemReporter(Consumer<LintProblem> problemConsumer) {
        this.problemConsumer = problemConsumer;
    }

    public void warn(NodeWithTokenRange<?> node, String category, String ruleId, String message, Object... args) {
        report(LintRule.WARN, node.getTokenRange().orElse(null), category, ruleId, message, args);
    }

    public void hint(NodeWithTokenRange<?> node, String category, String ruleId, String message, Object... args) {
        report(LintRule.HINT, node.getTokenRange().orElse(null), category, ruleId, message, args);
    }

    public void error(NodeWithTokenRange<?> node, String category, String ruleId, String message, Object... args) {
        report(LintRule.ERROR, node.getTokenRange().orElse(null), category, ruleId, message, args);
    }

    public void report(String level, TokenRange range, String category, String ruleId, String message, Object... args) {
        problemConsumer.accept(new LintProblem(level, f(message, args), range, null, category, ruleId));
    }

    public void report(LintProblem lintProblem) {
        problemConsumer.accept(lintProblem);
    }
}

File: tools/lint/src/main/kotlin/io/github/jmltoolkit/lint/LintProblemReporter.kt
package io.github.jmltoolkit.lint

import com.github.javaparser.TokenRange
import com.github.javaparser.ast.nodeTypes.NodeWithTokenRange
import java.util.function.Consumer

/**
 * @author Alexander Weigl
 * @version 1 (13.10.22)
 */
class LintProblemReporter(private val problemConsumer: Consumer<LintProblem>) {

    fun warn(node: NodeWithTokenRange<*>, category: String?, ruleId: String, message: String, vararg args: Any?) {
        report(LintRule.WARN, node.tokenRange.orElse(null), category, ruleId, message, args)
    }

    fun hint(node: NodeWithTokenRange<*>, category: String?, ruleId: String, message: String, vararg args: Any?) {
        report(LintRule.HINT, node.tokenRange.orElse(null), category, ruleId, message, args)
    }

    fun error(node: NodeWithTokenRange<*>, category: String?, ruleId: String, message: String, vararg args: Any?) {
        report(LintRule.ERROR, node.tokenRange.orElse(null), category, ruleId, message, args)
    }

    fun report(
        level: String,
        range: TokenRange?,
        category: String?,
        ruleId: String,
        message: String,
        vararg args: Any?
    ) {
        problemConsumer.accept(LintProblem(level, message.format(args), range, null, category, ruleId))
    }

    fun report(lintProblem: LintProblem) {
        problemConsumer.accept(lintProblem)
    }
}

